compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -600,6 +601,25 @@ object Semantic {
         || (meth eq defn.Object_ne)
         || (meth eq defn.Any_isInstanceOf)
 
+      def checkArgsWithParametricity() =
+        val methodType = atPhaseBeforeTransforms { meth.info.stripPoly }
+        var allArgsPromote = true
+        var singleParamSet = immutable.HashSet.from(methodType.paramInfoss.flatten.map(_.repeatedToSingle))
+        val errors = methodType.paramInfoss.flatten.zip(args).flatMap { (info, arg) =>
+          val singleInfo = info.repeatedToSingle
+          val isMatchable = defn.NothingType <:< singleInfo && singleInfo <:< defn.MatchableType
+          val isTypeParam = singleInfo.isInstanceOf[TypeParamRef]
+          val otherParamContains = singleParamSet.exists( param => { param != singleInfo && param.typeSymbol != defn.ClassTagClass && singleInfo.occursIn(param) } )
```

Comment:
Here, we only need to check other parameters if `isTypeParam` is true, and we need to check the bounds of the current type parameter.

The following is just a sketch for reference:

```Scala
val errors = arg.promote
allArgsPromote = allArgsPromote && errors.isEmpty

info match
case typeParamRef: TypeParamRef =>
  val bounds = typeParamRef.info.bounds
  def otherParamContains = allParamTypes.exists { param => 
    param != info && param.typeSymbol != defn.ClassTagClass && info.occursIn(param)
  }
  if bounds.lo <:< defn.NothingType && defn.AnyType <:< bounds.hi && !otherParamContains then Nil
  else errors
case _ =>
  errors
```

The reason we want to check the actual bounds is to avoid unsoundness when there are multiple type parameters, such as the following:

```
def foo[S, S <: T](x: S, arr: Array[T]) = arr(0) = x
```